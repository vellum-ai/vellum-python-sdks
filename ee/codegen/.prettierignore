# Ignore snapshots
*.ts.snap

# Ignore generated assets
src/assets/node-definitions.json
