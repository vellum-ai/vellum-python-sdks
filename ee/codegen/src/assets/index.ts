// This directory contains generated assets that are copied during postinstall
// and included in the npm package build. The node-definitions.json file is
// generated from the Python SDK and copied here for use in code generation.
