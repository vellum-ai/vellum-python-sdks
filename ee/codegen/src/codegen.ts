import {
  ErrorLogFile,
  InitFile,
  Inputs,
  NodeInput,
  VellumValue,
  VellumVariable,
  Workflow,
} from "./generators";
import { BasePersistedFile } from "./generators/base-persisted-file";
import { State } from "./generators/state";
import { WorkflowSandboxFile } from "./generators/workflow-sandbox-file";

export function vellumVariable(args: VellumVariable.Args): VellumVariable {
  return new VellumVariable(args);
}

export function vellumValue(args: VellumValue.Args): VellumValue {
  return new VellumValue(args);
}

export function nodeInput(args: NodeInput.Args): NodeInput {
  return new NodeInput(args);
}

export function inputs(args: Inputs.Args): Inputs {
  return new Inputs(args);
}

export function state(args: State.Args): State {
  return new State(args);
}

export function workflow(args: Workflow.Args): Workflow {
  return new Workflow(args);
}

export function initFile(args: InitFile.Args): InitFile {
  return new InitFile(args);
}

export function errorLogFile(args: BasePersistedFile.Args): ErrorLogFile {
  return new ErrorLogFile(args);
}

export function workflowSandboxFile(
  args: WorkflowSandboxFile.Args
): WorkflowSandboxFile {
  return new WorkflowSandboxFile(args);
}
