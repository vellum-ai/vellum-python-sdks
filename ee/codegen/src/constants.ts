/* Module paths */
export const VELLUM_CLIENT_MODULE_PATH = ["vellum"] as const;
export const VELLUM_WORKFLOW_NODE_BASE_TYPES_PATH = [
  "vellum",
  "workflows",
  "nodes",
  "displayable",
  "bases",
  "types",
] as const;
export const VELLUM_WORKFLOW_GRAPH_MODULE_PATH = [
  "vellum",
  "workflows",
  "graph",
] as const;
export const VELLUM_WORKFLOW_NODES_MODULE_PATH = [
  "vellum",
  "workflows",
  "nodes",
] as const;
export const VELLUM_WORKFLOW_CONSTANTS_PATH = [
  "vellum",
  "workflows",
  "references",
  "constant",
] as const;
export const VELLUM_WORKFLOWS_DISPLAY_BASE_PATH = [
  "vellum_ee",
  "workflows",
  "display",
  "base",
] as const;
export const VELLUM_WORKFLOW_EDITOR_TYPES_PATH = [
  "vellum_ee",
  "workflows",
  "display",
  "editor",
] as const;
/* Class names */
export const OUTPUTS_CLASS_NAME = "Outputs";
export const PORTS_CLASS_NAME = "Ports";
export const DEFAULT_PORT_NAME = "default";

/* File names */
export const INIT_FILE_NAME = "__init__.py";

export const GENERATED_WORKFLOW_MODULE_NAME = "workflow";

export const GENERATED_INPUTS_MODULE_NAME = "inputs";

export const GENERATED_NODES_MODULE_NAME = "nodes";

export const GENERATED_DISPLAY_MODULE_NAME = "display";
export const GENERATED_NODES_PATH = [GENERATED_NODES_MODULE_NAME] as const;

export const GENERATED_DISPLAY_NODE_MODULE_PATH = [
  GENERATED_DISPLAY_MODULE_NAME,
  GENERATED_NODES_MODULE_NAME,
] as const;

export const GENERATED_NESTED_NODE_MODULE_NAME = "node";
