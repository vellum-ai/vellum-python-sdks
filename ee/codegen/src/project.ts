import { mkdir } from "fs/promises";
import { join } from "path";

import { python } from "@fern-api/python-ast";
import { Comment } from "@fern-api/python-ast/Comment";
import { StarImport } from "@fern-api/python-ast/StarImport";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { VellumEnvironmentUrls } from "vellum-ai";

import * as codegen from "./codegen";
import {
  GENERATED_DISPLAY_MODULE_NAME,
  GENERATED_DISPLAY_NODE_MODULE_PATH,
  GENERATED_NODES_MODULE_NAME,
  GENERATED_NODES_PATH,
} from "./constants";
import { createNodeContext, WorkflowContext } from "./context";
import { InputVariableContext } from "./context/input-variable-context";
import { WorkflowOutputContext } from "./context/workflow-output-context";
import { ErrorLogFile, InitFile, Inputs, Workflow } from "./generators";
import {
  NodeDefinitionGenerationError,
  ProjectSerializationError,
} from "./generators/errors";
import { BaseNode } from "./generators/nodes/bases";
import { GuardrailNode } from "./generators/nodes/guardrail-node";
import { InlineSubworkflowNode } from "./generators/nodes/inline-subworkflow-node";
import { SearchNode } from "./generators/nodes/search-node";
import { TemplatingNode } from "./generators/nodes/templating-node";

import { ApiNodeContext } from "src/context/node-context/api-node";
import { BaseNodeContext } from "src/context/node-context/base";
import { CodeExecutionContext } from "src/context/node-context/code-execution-node";
import { ConditionalNodeContext } from "src/context/node-context/conditional-node";
import { ErrorNodeContext } from "src/context/node-context/error-node";
import { FinalOutputNodeContext } from "src/context/node-context/final-output-node";
import { GenericNodeContext } from "src/context/node-context/generic-node";
import { GuardrailNodeContext } from "src/context/node-context/guardrail-node";
import { InlinePromptNodeContext } from "src/context/node-context/inline-prompt-node";
import { InlineSubworkflowNodeContext } from "src/context/node-context/inline-subworkflow-node";
import { MapNodeContext } from "src/context/node-context/map-node";
import { MergeNodeContext } from "src/context/node-context/merge-node";
import { NoteNodeContext } from "src/context/node-context/note-node";
import { PromptDeploymentNodeContext } from "src/context/node-context/prompt-deployment-node";
import { SubworkflowDeploymentNodeContext } from "src/context/node-context/subworkflow-deployment-node";
import { TemplatingNodeContext } from "src/context/node-context/templating-node";
import { TextSearchNodeContext } from "src/context/node-context/text-search-node";
import { OutputVariableContext } from "src/context/output-variable-context";
import { ApiNode } from "src/generators/nodes/api-node";
import { CodeExecutionNode } from "src/generators/nodes/code-execution-node";
import { ConditionalNode } from "src/generators/nodes/conditional-node";
import { ErrorNode } from "src/generators/nodes/error-node";
import { FinalOutputNode } from "src/generators/nodes/final-output-node";
import { GenericNode } from "src/generators/nodes/generic-node";
import { InlinePromptNode } from "src/generators/nodes/inline-prompt-node";
import { MapNode } from "src/generators/nodes/map-node";
import { MergeNode } from "src/generators/nodes/merge-node";
import { NoteNode } from "src/generators/nodes/note-node";
import { PromptDeploymentNode } from "src/generators/nodes/prompt-deployment-node";
import { SubworkflowDeploymentNode } from "src/generators/nodes/subworkflow-deployment-node";
import { WorkflowSandboxFile } from "src/generators/workflow-sandbox-file";
import { WorkflowVersionExecConfigSerializer } from "src/serializers/vellum";
import {
  FinalOutputNode as FinalOutputNodeType,
  WorkflowDataNode,
  WorkflowNodeType as WorkflowNodeTypeEnum,
  WorkflowSandboxInputs,
  WorkflowVersionExecConfig,
} from "src/types/vellum";
import { assertUnreachable, isNilOrEmpty } from "src/utils/typing";

export interface WorkflowProjectGeneratorOptions {
  /**
   * Used to override the default codegen behavior for Code Execution Nodes. If set to "STANDALONE",
   *  the node's code will be generated in a separate file. If set to "INLINE", the node's code will
   *  be inlined as a node attribute.
   */
  codeExecutionNodeCodeRepresentationOverride?: "STANDALONE" | "INLINE";
  disableFormatting?: boolean;
}

export declare namespace WorkflowProjectGenerator {
  interface BaseArgs {
    moduleName: string;
    strict?: boolean;
  }

  interface BaseProject extends BaseArgs {
    absolutePathToOutputDirectory: string;
    workflowsSdkModulePath?: readonly string[];
    workflowVersionExecConfigData: unknown;
    vellumApiKey?: string;
    vellumApiEnvironment?: VellumEnvironmentUrls;
    sandboxInputs?: WorkflowSandboxInputs[];
    options?: WorkflowProjectGeneratorOptions;
  }

  interface NestedProject extends BaseArgs {
    workflowContext: WorkflowContext;
    workflowVersionExecConfig: WorkflowVersionExecConfig;
  }

  type Args = BaseProject | NestedProject;
}

export class WorkflowProjectGenerator {
  public readonly workflowVersionExecConfig: WorkflowVersionExecConfig;
  public readonly workflowContext: WorkflowContext;
  private readonly sandboxInputs?: WorkflowSandboxInputs[];

  constructor({ moduleName, ...rest }: WorkflowProjectGenerator.Args) {
    if ("workflowContext" in rest) {
      this.workflowContext = rest.workflowContext;
      this.workflowVersionExecConfig = rest.workflowVersionExecConfig;
    } else {
      const workflowVersionExecConfigResult =
        WorkflowVersionExecConfigSerializer.parse(
          rest.workflowVersionExecConfigData,
          {
            allowUnrecognizedUnionMembers: true,
            allowUnrecognizedEnumValues: true,
            unrecognizedObjectKeys: "strip",
          }
        );
      if (!workflowVersionExecConfigResult.ok) {
        const { errors } = workflowVersionExecConfigResult;
        if (errors.length) {
          throw new ProjectSerializationError(
            `Invalid Workflow Version exec config. Found ${
              errors.length
            } errors, including:
${errors.slice(0, 3).map((err) => {
  return `- ${err.message} at ${err.path.join(".")}`;
})}`
          );
        } else {
          throw new ProjectSerializationError(
            "Invalid workflow version exec config, but no errors were returned."
          );
        }
      }
      const vellumApiKey = rest.vellumApiKey ?? process.env.VELLUM_API_KEY;
      if (!vellumApiKey) {
        throw new ProjectSerializationError(
          "No workspace API key provided or found in environment variables."
        );
      }

      this.workflowVersionExecConfig = workflowVersionExecConfigResult.value;

      const workflowClassName =
        this.workflowVersionExecConfig.workflowRawData.definition?.name ||
        "Workflow";

      this.workflowContext = new WorkflowContext({
        workflowsSdkModulePath: rest.workflowsSdkModulePath,
        absolutePathToOutputDirectory: rest.absolutePathToOutputDirectory,
        moduleName,
        workflowClassName,
        vellumApiKey,
        vellumApiEnvironment: rest.vellumApiEnvironment,
        workflowRawData: this.workflowVersionExecConfig.workflowRawData,
        strict: rest.strict ?? false,
        disableFormatting: rest.options?.disableFormatting ?? false,
      });
      this.sandboxInputs = rest.sandboxInputs;
    }
  }

  public getModuleName(): string {
    return this.workflowContext.moduleName;
  }

  public async generateCode(): Promise<void> {
    const assets = await this.generateAssets().catch((error) => {
      this.workflowContext.addError(error);
      return null;
    });

    if (!assets) {
      return;
    }

    const { inputs, workflow, nodes } = assets;

    const absolutePathToModuleDirectory = join(
      this.workflowContext.absolutePathToOutputDirectory,
      this.workflowContext.moduleName
    );

    await mkdir(absolutePathToModuleDirectory, {
      recursive: true,
    });

    await Promise.all([
      // __init__.py
      this.generateRootInitFile().persist(),
      // display/__init__.py
      this.generateDisplayRootInitFile().persist(),
      // display/workflow.py
      workflow.getWorkflowDisplayFile().persist(),
      // inputs.py
      inputs.persist(),
      // workflow.py
      workflow.getWorkflowFile().persist(),
      // nodes/*
      ...this.generateNodeFiles(nodes),
      // sandbox.py
      ...(this.sandboxInputs ? [this.generateSandboxFile().persist()] : []),
    ]);

    // error.log - this gets generated separately from the other files because it
    // collects errors raised by the rest of the codegen process
    await this.generateErrorLogFile().persist();
  }

  private generateRootInitFile(): InitFile {
    const statements: AstNode[] = [];
    const imports: StarImport[] = [];
    const comments: Comment[] = [];

    const parentNode = this.workflowContext.parentNode;
    if (parentNode) {
      statements.push(parentNode.generateNodeClass());
    } else {
      comments.push(python.comment({ docs: "flake8: noqa: F401, F403" }));
      imports.push(
        python.starImport({
          modulePath: [
            this.workflowContext.moduleName,
            GENERATED_DISPLAY_MODULE_NAME,
          ],
        })
      );
    }

    const rootInitFile = codegen.initFile({
      workflowContext: this.workflowContext,
      modulePath: this.workflowContext.parentNode
        ? [...this.workflowContext.parentNode.getNodeModulePath()]
        : [this.workflowContext.moduleName],
      statements,
      imports,
      comments,
    });

    return rootInitFile;
  }

  private generateDisplayRootInitFile(): InitFile {
    const statements: AstNode[] = [];
    const imports: StarImport[] = [];
    const comments: Comment[] = [];

    const parentNode = this.workflowContext.parentNode;
    if (parentNode) {
      statements.push(...parentNode.generateNodeDisplayClasses());
      comments.push(python.comment({ docs: "flake8: noqa: F401, F403" }));
      imports.push(
        python.starImport({
          modulePath: [...parentNode.getNodeDisplayModulePath(), "nodes"],
        })
      );
      imports.push(
        python.starImport({
          modulePath: [...parentNode.getNodeDisplayModulePath(), "workflow"],
        })
      );
    } else {
      comments.push(python.comment({ docs: "flake8: noqa: F401, F403" }));
      imports.push(
        python.starImport({
          modulePath: [
            this.workflowContext.moduleName,
            GENERATED_DISPLAY_MODULE_NAME,
            "nodes",
          ],
        })
      );
      imports.push(
        python.starImport({
          modulePath: [
            this.workflowContext.moduleName,
            GENERATED_DISPLAY_MODULE_NAME,
            "workflow",
          ],
        })
      );
    }

    const rootDisplayInitFile = codegen.initFile({
      workflowContext: this.workflowContext,
      modulePath: this.workflowContext.parentNode
        ? [...this.workflowContext.parentNode.getNodeDisplayModulePath()]
        : [this.workflowContext.moduleName, GENERATED_DISPLAY_MODULE_NAME],
      statements,
      imports,
      comments,
    });

    return rootDisplayInitFile;
  }

  private async generateAssets(): Promise<{
    inputs: Inputs;
    workflow: Workflow;
    nodes: BaseNode<WorkflowDataNode, BaseNodeContext<WorkflowDataNode>>[];
  }> {
    const moduleName = this.workflowContext.moduleName;

    this.workflowVersionExecConfig.inputVariables.forEach((inputVariable) => {
      const inputVariableContext = new InputVariableContext({
        inputVariableData: inputVariable,
        workflowContext: this.workflowContext,
      });
      this.workflowContext.addInputVariableContext(inputVariableContext);
    });

    // TODO: Invert / remove this logic once output values are default and terminal nodes don't exist.
    // We are prioritizing terminal nodes as a temporary workaround to bad data from output values lingering in workflows.
    const terminalNodes =
      this.workflowVersionExecConfig.workflowRawData.nodes.filter(
        (nodeData): nodeData is FinalOutputNodeType =>
          nodeData.type === "TERMINAL"
      );

    if (terminalNodes.length > 0) {
      // Create from terminal nodes
      terminalNodes.forEach((nodeData) => {
        const outputVariableContext = new OutputVariableContext({
          outputVariableData: {
            id: nodeData.data.outputId,
            key: nodeData.data.name,
            type: nodeData.data.outputType,
          },
          workflowContext: this.workflowContext,
        });
        this.workflowContext.addOutputVariableContext(outputVariableContext);

        const workflowOutputContext = new WorkflowOutputContext({
          workflowContext: this.workflowContext,
          terminalNodeData: nodeData,
        });
        this.workflowContext.addWorkflowOutputContext(workflowOutputContext);
      });
    } else if (!isNilOrEmpty(this.workflowVersionExecConfig.outputVariables)) {
      const outputValuesById = Object.fromEntries(
        this.workflowVersionExecConfig.workflowRawData.outputValues?.map(
          (outputValue) => [outputValue.outputVariableId, outputValue]
        ) ?? []
      );
      this.workflowVersionExecConfig.outputVariables.forEach(
        (outputVariable) => {
          const outputVariableContext = new OutputVariableContext({
            outputVariableData: outputVariable,
            workflowContext: this.workflowContext,
          });
          this.workflowContext.addOutputVariableContext(outputVariableContext);

          const workflowOutput = outputValuesById[outputVariable.id];
          if (workflowOutput) {
            const workflowOutputContext = new WorkflowOutputContext({
              workflowContext: this.workflowContext,
              workflowOutputValue: workflowOutput,
            });
            this.workflowContext.addWorkflowOutputContext(
              workflowOutputContext
            );
          }
        }
      );
    }

    const nodesToGenerate = await Promise.all(
      this.getOrderedNodes().map(async (nodeData) => {
        await createNodeContext({
          workflowContext: this.workflowContext,
          nodeData,
        }).catch((error) => this.workflowContext.addError(error));
        return nodeData;
      })
    );

    const inputs = codegen.inputs({
      workflowContext: this.workflowContext,
    });

    const nodeIds = nodesToGenerate.map((nodeData) => nodeData.id);
    const nodes = await this.generateNodes(nodeIds);

    const workflow = codegen.workflow({
      moduleName,
      workflowContext: this.workflowContext,
      inputs,
      displayData: this.workflowVersionExecConfig.workflowRawData.displayData,
    });

    return { inputs, workflow, nodes };
  }

  /**
   * This method is used to order the nodes based on dependencies between them.
   * It ensures that nodes appear in the result after all the nodes they depend on.
   */
  private getOrderedNodes(): WorkflowDataNode[] {
    const rawData = this.workflowVersionExecConfig.workflowRawData;
    const nodes = rawData.nodes.filter(
      (node): node is WorkflowDataNode => node.type !== "ENTRYPOINT"
    );

    // Edge case: Workflow init only has two nodes of ENTRYPOINT and TERMINAL with no edge between them
    if (rawData.edges.length === 0) {
      return nodes;
    }

    const nodesById = Object.fromEntries(nodes.map((node) => [node.id, node]));

    // Create a dependency graph that represents the "depends on" relationships between nodes
    // If A depends on B, then B must come before A in the result
    const dependencyGraph: Map<string, Set<string>> = new Map();

    // Initialize dependency graph with all nodes
    nodes.forEach((node) => {
      dependencyGraph.set(node.id, new Set());
    });

    // Build dependency relationships based on edges
    // Process source node before target node
    rawData.edges.forEach((edge) => {
      dependencyGraph.get(edge.targetNodeId)?.add(edge.sourceNodeId);
    });

    const orderedNodes: WorkflowDataNode[] = [];
    const visited = new Set<string>();
    const tempVisited = new Set<string>(); // Used to detect cycles

    // DFS function that ensures dependencies are processed first
    const visit = (nodeId: string) => {
      // Skip if already processed
      if (visited.has(nodeId)) {
        return;
      }

      // Return early if we've already visited this node in the current path (cycle)
      if (tempVisited.has(nodeId)) {
        return;
      }

      // Mark as temporarily visited (part of current path)
      tempVisited.add(nodeId);

      // Process all dependencies first
      const dependencies = dependencyGraph.get(nodeId) || new Set();
      for (const depId of dependencies) {
        visit(depId);
      }

      // All dependencies processed, now safe to add this node
      const node = nodesById[nodeId];
      if (node) {
        orderedNodes.push(node);
      }

      // Mark as fully processed
      visited.add(nodeId);
      tempVisited.delete(nodeId);
    };

    // Visit each node to ensure all are processed
    nodes.forEach((node) => {
      if (!visited.has(node.id)) {
        visit(node.id);
      }
    });

    return orderedNodes;
  }

  private async generateNodes(
    nodeIds: string[]
  ): Promise<BaseNode<WorkflowDataNode, BaseNodeContext<WorkflowDataNode>>[]> {
    const nodes: BaseNode<
      WorkflowDataNode,
      BaseNodeContext<WorkflowDataNode>
    >[] = [];

    await Promise.all(
      nodeIds.map(async (nodeId) => {
        let node: BaseNode<WorkflowDataNode, BaseNodeContext<WorkflowDataNode>>;

        const nodeContext = this.workflowContext.findNodeContext(nodeId);
        if (!nodeContext) {
          return;
        }

        const nodeData = nodeContext.nodeData;

        const nodeType = nodeData.type;
        switch (nodeType) {
          case WorkflowNodeTypeEnum.SEARCH: {
            node = new SearchNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as TextSearchNodeContext,
            });
            break;
          }
          case WorkflowNodeTypeEnum.SUBWORKFLOW: {
            const variant = nodeData.data.variant;
            switch (variant) {
              case "INLINE":
                node = new InlineSubworkflowNode({
                  workflowContext: this.workflowContext,
                  nodeContext: nodeContext as InlineSubworkflowNodeContext,
                });
                break;
              case "DEPLOYMENT":
                node = new SubworkflowDeploymentNode({
                  workflowContext: this.workflowContext,
                  nodeContext: nodeContext as SubworkflowDeploymentNodeContext,
                });
                break;
              default: {
                assertUnreachable(variant);
              }
            }
            break;
          }
          case WorkflowNodeTypeEnum.MAP: {
            const mapNodeVariant = nodeData.data.variant;
            switch (mapNodeVariant) {
              case "INLINE":
                node = new MapNode({
                  workflowContext: this.workflowContext,
                  nodeContext: nodeContext as MapNodeContext,
                });
                break;
              case "DEPLOYMENT":
                throw new NodeDefinitionGenerationError(
                  `DEPLOYMENT variant not yet supported`
                );
              default: {
                assertUnreachable(mapNodeVariant);
              }
            }
            break;
          }
          case WorkflowNodeTypeEnum.METRIC: {
            node = new GuardrailNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as GuardrailNodeContext,
            });
            break;
          }
          case WorkflowNodeTypeEnum.CODE_EXECUTION: {
            node = new CodeExecutionNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as CodeExecutionContext,
            });
            break;
          }
          case WorkflowNodeTypeEnum.PROMPT: {
            const promptNodeVariant = nodeData.data.variant;

            switch (promptNodeVariant) {
              case "INLINE":
                node = new InlinePromptNode({
                  workflowContext: this.workflowContext,
                  nodeContext: nodeContext as InlinePromptNodeContext,
                });
                break;
              case "DEPLOYMENT":
                node = new PromptDeploymentNode({
                  workflowContext: this.workflowContext,
                  nodeContext: nodeContext as PromptDeploymentNodeContext,
                });
                break;
              case "LEGACY":
                throw new NodeDefinitionGenerationError(
                  `LEGACY variant should have been converted to INLINE variant by this point.`
                );
              default: {
                assertUnreachable(promptNodeVariant);
              }
            }
            break;
          }
          case WorkflowNodeTypeEnum.TEMPLATING: {
            node = new TemplatingNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as TemplatingNodeContext,
            });
            break;
          }
          case WorkflowNodeTypeEnum.CONDITIONAL: {
            node = new ConditionalNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as ConditionalNodeContext,
            });
            break;
          }
          case WorkflowNodeTypeEnum.TERMINAL: {
            node = new FinalOutputNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as FinalOutputNodeContext,
            });
            break;
          }
          case WorkflowNodeTypeEnum.MERGE: {
            node = new MergeNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as MergeNodeContext,
            });
            break;
          }
          case WorkflowNodeTypeEnum.ERROR: {
            node = new ErrorNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as ErrorNodeContext,
            });
            break;
          }
          case WorkflowNodeTypeEnum.NOTE: {
            node = new NoteNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as NoteNodeContext,
            });
            break;
          }
          case WorkflowNodeTypeEnum.API:
            node = new ApiNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as ApiNodeContext,
            });
            break;
          case WorkflowNodeTypeEnum.GENERIC:
            node = new GenericNode({
              workflowContext: this.workflowContext,
              nodeContext: nodeContext as GenericNodeContext,
            });
            break;
          default: {
            throw new NodeDefinitionGenerationError(
              `Unsupported node type: ${nodeType}`
            );
          }
        }

        nodes.push(node);
      })
    );

    return nodes;
  }

  private sortAlphabetically(items: string[]): string[] {
    return items.sort((a, b) => a.localeCompare(b));
  }

  private generateNodeFiles(
    nodes: BaseNode<WorkflowDataNode, BaseNodeContext<WorkflowDataNode>>[]
  ): Promise<unknown>[] {
    const rootNodesInitFileStatements: AstNode[] = [];
    const rootDisplayNodesInitFileStatements: AstNode[] = [];
    if (nodes.length) {
      const nodeInitFileAllField = python.field({
        name: "__all__",
        initializer: python.TypeInstantiation.list(
          [
            ...this.sortAlphabetically(
              nodes.map((node) => node.getNodeClassName())
            ).map((name) => python.TypeInstantiation.str(name)),
          ],
          {
            endWithComma: true,
          }
        ),
      });
      rootNodesInitFileStatements.push(nodeInitFileAllField);

      const nodeDisplayInitFileAllField = python.field({
        name: "__all__",
        initializer: python.TypeInstantiation.list(
          [
            ...this.sortAlphabetically(
              nodes.map((node) => node.getNodeDisplayClassName())
            ).map((name) => python.TypeInstantiation.str(name)),
          ],
          {
            endWithComma: true,
          }
        ),
      });
      rootDisplayNodesInitFileStatements.push(nodeDisplayInitFileAllField);
    }

    const rootNodesInitFile = codegen.initFile({
      workflowContext: this.workflowContext,
      modulePath: this.workflowContext.parentNode
        ? [
            ...this.workflowContext.parentNode.getNodeModulePath(),
            GENERATED_NODES_MODULE_NAME,
          ]
        : [this.workflowContext.moduleName, ...GENERATED_NODES_PATH],
      statements: rootNodesInitFileStatements,
    });

    const rootDisplayNodesInitFile = codegen.initFile({
      workflowContext: this.workflowContext,
      modulePath: this.workflowContext.parentNode
        ? [
            ...this.workflowContext.parentNode.getNodeDisplayModulePath(),
            GENERATED_NODES_MODULE_NAME,
          ]
        : [
            this.workflowContext.moduleName,
            ...GENERATED_DISPLAY_NODE_MODULE_PATH,
          ],
      statements: rootDisplayNodesInitFileStatements,
    });

    nodes.forEach((node) => {
      rootNodesInitFile.addReference(
        python.reference({
          name: node.getNodeClassName(),
          modulePath: node.getNodeModulePath(),
        })
      );

      rootDisplayNodesInitFile.addReference(
        python.reference({
          name: node.getNodeDisplayClassName(),
          modulePath: node.getNodeDisplayModulePath(),
        })
      );
    });

    const nodePromises = nodes.map(async (node) => {
      return await node.persist();
    });

    return [
      // nodes/__init__.py
      rootNodesInitFile.persist(),
      // display/nodes/__init__.py
      rootDisplayNodesInitFile.persist(),
      // nodes/* and display/nodes/*
      ...nodePromises,
    ];
  }

  private generateErrorLogFile(): ErrorLogFile {
    return codegen.errorLogFile({
      workflowContext: this.workflowContext,
    });
  }

  private generateSandboxFile(): WorkflowSandboxFile {
    return codegen.workflowSandboxFile({
      workflowContext: this.workflowContext,
      sandboxInputs: this.sandboxInputs ?? [],
    });
  }

  /**
   * Gets the node file paths that have been tracked during code generation that will be merged by codegen-service
   * @returns Set of node file paths
   */
  public getPythonCodeMergeableNodeFiles(): Set<string> {
    return this.workflowContext.getPythonCodeMergeableNodeFiles();
  }
}
