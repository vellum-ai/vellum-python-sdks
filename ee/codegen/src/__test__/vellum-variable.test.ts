import { Writer } from "@fern-api/python-ast/core/Writer";

import * as codegen from "src/codegen";

describe("VellumVariableField", () => {
  let writer: Writer;

  beforeEach(() => {
    writer = new Writer();
  });

  test("StringVellumVariable snapshot", async () => {
    const stringVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "STRING", required: true },
    });
    stringVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("NumberVellumVariable snapshot", async () => {
    const numberVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "NUMBER", required: true },
    });
    numberVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("JsonVellumVariable snapshot", async () => {
    const jsonVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "JSON", required: true },
    });
    jsonVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("ImageVellumVariable snapshot", async () => {
    const imageVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "IMAGE", required: true },
    });
    imageVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("DocumentVellumVariable snapshot", async () => {
    const documentVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "DOCUMENT", required: true },
    });
    documentVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("FunctionCallVellumVariable snapshot", async () => {
    const functionCallVar = codegen.vellumVariable({
      variable: {
        id: "1",
        name: "test",
        type: "FUNCTION_CALL",
        required: true,
      },
    });
    functionCallVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("ErrorVellumVariable snapshot", async () => {
    const errorVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "ERROR", required: true },
    });
    errorVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("ArrayVellumVariable snapshot", async () => {
    const arrayVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "ARRAY", required: true },
    });
    arrayVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("ChatHistoryVellumVariable snapshot", async () => {
    const chatHistoryVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "CHAT_HISTORY", required: true },
    });
    chatHistoryVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("SearchResultsVellumVariable snapshot", async () => {
    const searchResultsVar = codegen.vellumVariable({
      variable: {
        id: "1",
        name: "test",
        type: "SEARCH_RESULTS",
        required: true,
      },
    });
    searchResultsVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("NullVellumVariable snapshot", async () => {
    const nullVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "NULL" },
    });
    nullVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("OptionalVellumVariable defaultRequired=true snapshot", async () => {
    const optionalVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "STRING" },
      defaultRequired: true,
    });
    optionalVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  test("OptionalVellumVariable defaultRequired=false snapshot", async () => {
    const optionalVar = codegen.vellumVariable({
      variable: { id: "1", name: "test", type: "STRING" },
      defaultRequired: false,
    });
    optionalVar.write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});
