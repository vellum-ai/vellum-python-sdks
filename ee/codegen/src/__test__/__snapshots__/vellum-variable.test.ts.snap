// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VellumVariableField > ArrayVellumVariable snapshot 1`] = `
"test: List[VellumValue]
"
`;

exports[`VellumVariableField > ChatHistoryVellumVariable snapshot 1`] = `
"test: List[ChatMessage]
"
`;

exports[`VellumVariableField > DocumentVellumVariable snapshot 1`] = `
"test: VellumDocument
"
`;

exports[`VellumVariableField > ErrorVellumVariable snapshot 1`] = `
"test: VellumError
"
`;

exports[`VellumVariableField > FunctionCallVellumVariable snapshot 1`] = `
"test: FunctionCall
"
`;

exports[`VellumVariableField > ImageVellumVariable snapshot 1`] = `
"test: VellumImage
"
`;

exports[`VellumVariableField > JsonVellumVariable snapshot 1`] = `
"test: Any
"
`;

exports[`VellumVariableField > NullVellumVariable snapshot 1`] = `
"test: None
"
`;

exports[`VellumVariableField > NumberVellumVariable snapshot 1`] = `
"test: Union[float, int]
"
`;

exports[`VellumVariableField > OptionalVellumVariable defaultRequired=false snapshot 1`] = `
"test: Optional[str] = None
"
`;

exports[`VellumVariableField > OptionalVellumVariable defaultRequired=true snapshot 1`] = `
"test: str
"
`;

exports[`VellumVariableField > SearchResultsVellumVariable snapshot 1`] = `
"test: List[SearchResult]
"
`;

exports[`VellumVariableField > StringVellumVariable snapshot 1`] = `
"test: str
"
`;
